From 91d217e995941d35ea4583ba029d70c89e114ba6 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Mon, 19 Oct 2020 21:28:36 +0800
Subject: [PATCH] PONRTSYS-8082:GSWAPI: Fix for meter deletion

Meter index should be released assigned previously,when its disabled or not assigned
in bridge port and ctp port.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 61 ++++++++++++++++++++++
 1 file changed, 61 insertions(+)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index 0c0b68d5340f..d16904d99c3f 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -22587,6 +22587,8 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 	u16 i, val_reg_idx, ret, IngressStpState, EgressStpState;
 	u32 BlkSize = 0;
 	int retry;
+	gsw_brdgportconfig_t *brdge_pcfg =
+	&gswdev->brdgeportconfig_idx[param->nBridgePortId];
 
 	if (gswdev == NULL) {
 		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
@@ -23075,6 +23077,20 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Ingress Meter*/
 			tbl_prog_brdgeport_ingress.val[3] &= ~(1 << 7);
+			/** when Metering operation is disabled,it is must release
+			 * the meter idx if it is assigned previously.
+			 */
+			if (brdge_pcfg->IngressMeteringAssigned) {
+				idx = brdge_pcfg->IngressTrafficMeterId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->IngressMeteringAssigned = 0;
+				brdge_pcfg->IngressTrafficMeterId = 0;
+			}
 		}
 	}
 
@@ -23447,6 +23463,20 @@ GSW_return_t GSW_BridgePortConfigSet(void *cdev, GSW_BRIDGE_portConfig_t *param)
 		} else {
 			/*Disable Egress Meter*/
 			tbl_prog_brdgeport_egress.val[3] &= ~(1 << 7);
+			/** when Metering operation is disabled,it is must release
+			 * the meter idx if it is assigned previously
+			 */
+			if (brdge_pcfg->EgressMeteringAssigned) {
+				idx = brdge_pcfg->EgressTrafficMeterId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				brdge_pcfg->EgressMeteringAssigned = 0;
+				brdge_pcfg->EgressTrafficMeterId = 0;
+			}
 		}
 	}
 
@@ -24766,6 +24796,7 @@ GSW_return_t GSW_CtpPortConfigSet(void *cdev, GSW_CTP_portConfig_t *param)
 	u32 BlkSize = 0;
 	u32 ret;
 	GSW_CTP_portAssignment_t ctp_get;
+	gsw_ctpportconfig_t *ctp_pcfg;
 
 	if (gswdev == NULL) {
 		pr_err("%s:%s:%d", __FILE__, __func__, __LINE__);
@@ -24815,6 +24846,8 @@ GSW_return_t GSW_CtpPortConfigSet(void *cdev, GSW_CTP_portConfig_t *param)
 		goto UNLOCK_AND_RETURN;
 	}
 
+	ctp_pcfg = &gswdev->ctpportconfig_idx[ctp_port];
+
 	/*NOTE : This Mask is for debugging purpose only
 	  If GSW_CTP_PORT_CONFIG_MASK_FORCE mask is set
 	  It will not check index in-use */
@@ -25428,6 +25461,20 @@ GSW_return_t GSW_CtpPortConfigSet(void *cdev, GSW_CTP_portConfig_t *param)
 		} else {
 			/*Disable Ingress Meter*/
 			tbl_prog_ctpport_ingress.val[3] &= ~(1 << 7);
+			/** when Metering operation is disabled,it is must release
+			 * the meter idx if it is assigned previously.
+			 */
+			if (ctp_pcfg->IngressMeteringAssigned) {
+				idx = ctp_pcfg->IngressTrafficMeterId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+			gswdev->meter_idx[idx].IndexInUsageCnt--;
+			ctp_pcfg->IngressMeteringAssigned = 0;
+			ctp_pcfg->IngressTrafficMeterId = 0;
+			}
 		}
 	}
 
@@ -25488,6 +25535,20 @@ GSW_return_t GSW_CtpPortConfigSet(void *cdev, GSW_CTP_portConfig_t *param)
 		} else {
 			/*Disable Egress Meter*/
 			tbl_prog_ctpport_egress.val[3] &= ~(1 << 7);
+			/** when Metering operation is disabled,it is must release
+			 * the meter idx if it is assigned previously
+			 */
+			if (ctp_pcfg->EgressMeteringAssigned) {
+				idx = ctp_pcfg->EgressTrafficMeterId;
+				if (idx >= ARRAY_SIZE(gswdev->meter_idx)) {
+					pr_err("ERROR :idx for meter_idx %d out of bounds!", idx);
+					ret = GSW_statusErr;
+					goto UNLOCK_AND_RETURN;
+				}
+				gswdev->meter_idx[idx].IndexInUsageCnt--;
+				ctp_pcfg->EgressMeteringAssigned = 0;
+				ctp_pcfg->EgressTrafficMeterId = 0;
+			}
 		}
 	}
 
